diff --git a/lib/spack/env/cc b/lib/spack/env/cc
index 6ce60a8730..cfc2d4dca7 100755
--- a/lib/spack/env/cc
+++ b/lib/spack/env/cc
@@ -749,6 +749,21 @@ extend args_list other_args_list
 # Inject SPACK_LDLIBS, if supplied
 extend args_list libs_list "-l"
 
+# When using clang add some flags
+case "${0##*/}" in
+    clang|clang++|flang)
+        case "$mode" in
+            cc)
+                extend args_list SPACK_EXTRA_CC_FLAGS
+                ;;
+            ccld)
+                extend args_list SPACK_EXTRA_CC_FLAGS
+                extend args_list SPACK_EXTRA_CCLD_FLAGS
+                ;;
+        esac
+        ;;
+esac
+
 full_command_list="$command"
 extend full_command_list args_list
 
