From 6e09b8fb5f3be6ebc543789d11be916d63b0ab0b Mon Sep 17 00:00:00 2001
From: Harmen Stoppels <harmenstoppels@gmail.com>
Date: Wed, 25 May 2022 12:04:00 +0200
Subject: [PATCH] compiler wrapper hack

---
 lib/spack/env/cc | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/lib/spack/env/cc b/lib/spack/env/cc
index bef7209bfa..423f886ae7 100755
--- a/lib/spack/env/cc
+++ b/lib/spack/env/cc
@@ -725,6 +725,23 @@ extend args_list other_args_list
 # Inject SPACK_LDLIBS, if supplied
 extend args_list libs_list "-l"
 
+# When using clang or clang++ force lld, use -O3 in the linker, add
+case "${0##*/}" in
+    clang|clang++)
+        case "$mode" in
+            cc|ccld)
+               extend args_list SPACK_EXTRA_CFLAGS
+                ;;
+        esac
+
+        if [ "$mode" = "ccld" ]; then
+            append args_list "-fuse-ld=lld"
+            append args_list "-flto=thin"
+            append args_list "-Wl,-O3"
+        fi
+        ;;
+esac
+
 full_command_list="$command"
 extend full_command_list args_list
 
-- 
2.25.1

